src/OpenTelemetry/Metrics/IDataPoint.cs

```diff
@@ -0,0 +1,37 @@
+// <copyright file="IDataPoint.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics
+{
+    public interface IDataPoint
+    {
+        public DateTimeOffset Timestamp { get; }
+
+        public KeyValuePair<string, object>[] Tags { get; }
```

Comment:
Is there a good reason to have Tags be an array as opposed to something like IList or IEnumerable? Only because it will probably lead to .ToArrays() we might be able to avoid.

Reply:
When walking through an IEnumerable<>, an allocation occurs. Trying to avoid any allocation in hot path.

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.202
  [Host]     : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT
  DefaultJob : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT


|         Method | WithSDK |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |-------- |----------:|----------:|----------:|-------:|------:|------:|----------:|
|       WalkEnum |    True | 52.351 ns | 0.9519 ns | 1.0962 ns | 0.0076 |     - |     - |      32 B |
|      WalkArray |    True |  8.349 ns | 0.1109 ns | 0.0926 ns |      - |     - |     - |         - |
| WalkArrayIndex |    True |  9.557 ns | 0.2159 ns | 0.2807 ns |      - |     - |     - |         - |
```

``` c#
namespace Benchmarks.Metrics
{
    [MemoryDiagnoser]
    public class MetricsBenchmarks
    {
        private readonly KeyValuePair<string, object> tag1 = new KeyValuePair<string, object>("attrib1", "value1");
        private readonly KeyValuePair<string, object> tag2 = new KeyValuePair<string, object>("attrib2", "value2");
        private readonly KeyValuePair<string, object> tag3 = new KeyValuePair<string, object>("attrib3", "value3");
        private readonly KeyValuePair<string, object> tag4 = new KeyValuePair<string, object>("attrib4", "value4");
        private readonly KeyValuePair<string, object> tag5 = new KeyValuePair<string, object>("attrib5", "value5");

        private IEnumerable<KeyValuePair<string, object>> enumTags;
        private KeyValuePair<string, object>[] arrayTags;

        [GlobalSetup]
        public void Setup()
        {
            this.arrayTags = new KeyValuePair<string, object>[]
            {
                this.tag1,
                this.tag2,
                this.tag3,
                this.tag4,
                this.tag5,
            };

            this.enumTags = this.arrayTags;
        }

        [Benchmark]
        public void WalkEnum()
        {
            foreach (var tag in this.enumTags)
            {
                if (tag.Key == "Hello")
                {
                    throw new System.Exception();
                }
            }
        }

        [Benchmark]
        public void WalkArray()
        {
            foreach (var tag in this.arrayTags)
            {
                if (tag.Key == "Hello")
                {
                    throw new System.Exception();
                }
            }
        }

        [Benchmark]
        public void WalkArrayIndex()
        {
            for (int i = 0; i < this.arrayTags.Length; i++)
            {
                var tag = this.arrayTags[i];

                if (tag.Key == "Hello")
                {
                    throw new System.Exception();
                }
            }
        }
    }
}
```

Reply:
For sure. You can do the same thing (for loop) with IList<T> though. I think if we are going to pick a concrete type, List<T> might be more useful. You can foreach those without allocation, and I think they are generally easier to work with over array where you need to know your size going in.

Do these tags on metrics have the same behaviors as span tags? Like if you add null, is that a removal? If that is the case, we might want to use ActivityTagsCollection from runtime?

Reply:
I'll check with .NET team on maybe reusing ActivityTagsCollection.

Reply:
So, taking to @tarekgh from .NET team, ActivityTagsCollection is backed with a List<> object. It's performance profile will be equivalent to a List<>.

Doing some benchmarks, I make the following conclusion.

- Using IList<> will wind up with an alloc (presumably to get the Enumerator)
- Using List<> directly is alloc free, but cost 10x more cycles for simple foreach().

### Conclusion: Unless we need have frequent add/remove, I would recommend we stay with arrays. At this time, I don't think we need to support easy add/remove.

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.202
  [Host]     : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT
  DefaultJob : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT


|         Method | WithSDK |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |-------- |----------:|----------:|----------:|-------:|------:|------:|----------:|
|       WalkEnum |    True | 52.406 ns | 1.0599 ns | 0.9914 ns | 0.0076 |     - |     - |      32 B |
|      WalkArray |    True |  7.980 ns | 0.0918 ns | 0.0813 ns |      - |     - |     - |         - |
| WalkArrayIndex |    True |  9.451 ns | 0.2129 ns | 0.2278 ns |      - |     - |     - |         - |
|       WalkList |    True | 42.717 ns | 0.4898 ns | 0.4342 ns |      - |     - |     - |         - |
|      WalkIList |    True | 88.783 ns | 1.6342 ns | 1.4487 ns | 0.0114 |     - |     - |      48 B |
```

Reply:
Let me clarify my points a bit...

* Re: ActivityTagsCollection 

I doubt very much `KeyValuePair<string, object>[] Tags` is compliant with the spec 😄 I haven't looked in a while, but the spec has specific rules for tags, right? Setting null removes tag, setting the same key overrides any existing tag. Stuff like that. The array is not a smart enough type to encapsulate these rules, so user will have to understand them. ActivityTagsCollection has them already all built-in ready to go, so I was suggesting we piggyback on that. It will take care of the spec details so the user doesn't have to worry about them. Also has a struct enumerator so you can foreach it without allocation.

* Re: Perf of array

```csharp
public void WalkArray(KeyValuePair<string, object>[] tags)
{
   // Allocates an enumerator
   foreach (KeyValuePair<string, object> tag in tags) {}

   // Does not allocate
   for (int i = 0; i < tags.Length; i++)
   {
      KeyValuePair<string, object> tag = tags[i];
   }
}

public void WalkList(IList<KeyValuePair<string, object>> tags)
{
   // Allocates an enumerator
   foreach (KeyValuePair<string, object> tag in tags) {}

   // Does not allocate
   for (int i = 0; i < tags.Count; i++) 
   {
      KeyValuePair<string, object> tag = tags[i];
   }
}
```

Perf is more or less the same, right? The difference is, IList is less restrictive than array. Users have more options as far as what concrete type they use, basically.

Reply:
I opened #2062 to track and further discuss this feedback.